### PR TITLE
Master mass mailing segmentation mub

### PR DESCRIPTION
**Pull request by mub-odoo**
Whether targeting business objects or using mailing lists and contacts, marketing users usually refine the audience they're going to reach with a specific mailing thanks to filters.
This commit allow users to save filters in mass_mailing to be able to reuse mailing lists and not be forced to recreate the audience they want to target.

taskid:2092853

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by awa-odoo — COMMENTED** [review]
Hello @mub-odoo 

That new version with only 1 field is a lot better, thanks!
I added some polishing remarks, mostly about naming, but overall it's looking good.

Cheers!

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    },
+
+    /**
+     * Updates the 'Add to favorite' / 'Remove' star icons' visibility based on
+     * the current state.
+     *
+     * @private
+     */
+    _updateFilterIcons: function () {
+        const hasDomain = this.recordData.mailing_domain.length > 2;
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:61]: if no domain is applied than it's minimum length is 2
https://www.awesomescreenshot.com/image/12385128?key=144ae5c23f9982090a4092208053e7b5

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        // if the domain is empty, hide everything related to filters
+        this.el.querySelector('.o_mass_mailing_filter_container').classList.toggle('invisible', !hasDomain);
+        const canAddFilter = !this.recordData.filter_id || !this.$input.val().length || this.floating;
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:64]: Done !

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        this.el.querySelector('.o_mass_mailing_add_filter_container').classList.toggle('d-none', !canAddFilter);
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:65]: Done !

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    /**
+     * Deletes the saved filter, but we do not reset the applied domain
+     * in this case.
+     *
+     * @private
+     * @param {event} ev
+     */
+    _onRemoveClick(ev) {
+        this._rpc({
+            model: this.field.relation,
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:134]: Done !

File: addons/mass_mailing/static/src/xml/mass_mailing.xml [hunk 5]
```diff
@@ -15,4 +15,27 @@
...
+    <t t-name="mass_mailing.favorite_filter" t-extend="FieldMany2One" primary="True">
+        <t t-jquery="div.o_field_many2one_selection" t-operation="append">
+            <div class="o_mass_mailing_filter_container">
+                <div class="o_mass_mailing_add_filter_container">
+                    <i class="o_mass_mailing_add_filter fa fa-star-o px-3 py-1" data-toggle="dropdown" title="Add to Favorites"/>
+                    <div class="dropdown-menu dropdown-menu-right">
+                        <span class="ml-2 text-muted">Create a new Favorite Domain</span><br/>
+                        <div class="d-inline-flex">
+                            <input type="text" class="o_mass_mailing_filter_input w-auto ml-2" placeholder="E.g.: VIP customers" />
+                            <button tabindex="-1" class="btn btn-sm btn-primary o_mass_mailing_btn_save_filter mx-2">
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/xml/mass_mailing.xml:29]: Done !

File: addons/mass_mailing/tests/test_mailing_internals.py [hunk 6]
```diff
@@ -107,6 +107,54 @@ def test_mailing_computed_fields(self):
...
+        # changing the filter should update the mailing domain correctly
+        mailing.filter_id = favorite_filter_channel_2
+        self.assertEqual(literal_eval(mailing.mailing_domain), literal_eval(favorite_filter_channel_2.mailing_domain))
+
+        # changing the model should empty the filter_id, and set the default domain for the model
+        mailing.mailing_model_id = self.env['ir.model']._get('res.partner').id
+        self.assertFalse(mailing.filter_id, "should unset filter if model is changed")
+        self.assertEqual(literal_eval(mailing.mailing_domain), [('is_blacklisted', '=', False)], "default domain for the changed model is set")
+
+
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/tests/test_mailing_internals.py:143]: Done !

File: addons/mass_mailing/views/mailing_filters_views.xml [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <record id="mailing_filters_view_tree" model="ir.ui.view">
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/views/mailing_filters_views.xml:3]: in search more 
https://www.awesomescreenshot.com/image/12386396?key=439d9c3fa2147364936749bc3495e163

File: addons/mass_mailing/models/mailing.py [hunk 8]
```diff
@@ -124,6 +124,9 @@ def _get_default_mail_server_id(self):
     mailing_domain = fields.Char(
         string='Domain', compute='_compute_mailing_domain',
         readonly=False, store=True)
+    filter_id = fields.Many2one('mailing.filters', string='Filters',
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/models/mailing.py:127]: Done !

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+ * field on 'mailing.mailing' from view.
+ *
+ * In edit mode, it will allow to save the latest configured domain
+ * in form of the favorite filter, or to remove the store filters.
+ *
+ * @class
+ */
+const MassMailingFavoriteFilter = FieldMany2One.extend({
+    template: 'mass_mailing.favorite_filter',
+    events: _.extend({}, FieldMany2One.prototype.events, {
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:21]: Done !

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+        'click .o_mass_mailing_btn_save_filter': '_onSaveClick',
+        'click .o_mass_mailing_remove_filter': '_onRemoveClick',
+        'keydown .o_mass_mailing_filter_input': '_onFilterInputKeydown',
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:24]: Done !

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+        const filterInput = this.el.querySelector('input.o_mass_mailing_filter_input');
+        const filterName = filterInput.value.trim();
+        if (filterName.length === 0) {
+            this.displayNotification({ message: _t("Please provide name for the filter"), type: 'danger' });
+            // Keep the drop-down open, and re-focus the input
+            ev.stopPropagation();
+            filterInput.focus();
+        } else {
+            this._rpc({
+                model: this.field.relation,
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:171]: Done !

File: addons/mass_mailing/static/tests/mass_mailing_favourite_filter_tests.js [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+                        type: "char"
+                    },
+                    mailing_domain: {
+                        string: "Display name",
+                        type: "char"
+                    },
+                    mailing_model_name: {
+                        string: "Display name",
+                        type: "char"
+                    }
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/tests/mass_mailing_favourite_filter_tests.js:75]: Done !

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:61]: I see, would it be the same as doing: 

`const hasDomain = this.recordData.mailing_domain && this.recordData.mailing_domain !== '[]';`

I feel like it is more explicit.

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+import { FieldMany2One } from 'web.relational_fields';
+import core from 'web.core';
+import fieldRegistry from 'web.field_registry';
+
+const _t = core._t;
+
+/**
+ * Widget to create / remove favorite filters on mass mailing, extended
+ * from FieldMany2One. This widget is designed specifically for 'filter_id'
+ * field on 'mailing.mailing' from view.
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:12]: Done !

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:61]: Done !

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+     * This widget now has two inputs, one for m2o and another for filter
+     * name. So this override makes sure that 'this.$input' now refers to
+     * only main input instead of both.
+     *
+     * @override
+     */
+    start() {
+        let def = this._super.apply(this, arguments);
+        this.$input = this.$('input.o_input');
+        return def;
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:39]: Hello @awa-odoo 
I tried but got this issue:
https://youtu.be/enB6qsybOos
it consider both textbox as m20 input expect first one

File: addons/mass_mailing/static/tests/mass_mailing_favourite_filter_tests.js [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+                    mailing_domain: "[['country','=','be']]",
+                    mailing_model_name: "partner",
+                }],
+            }
+        };
+    },
+
+}, function () {
+
+    QUnit.test('create favorite filter', async function (assert) {
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/tests/mass_mailing_favourite_filter_tests.js:93]: Hello @awa-odoo 
I have added one test case. It checks that the onchange correctly applies the filter.

**Review by awa-odoo — COMMENTED** [review]
Hello @mub-odoo 

I had a more in-depth look.
Left some questions and suggestions.

Thanks for your work!

File: addons/mass_mailing/models/mailing.py [hunk 16]
```diff
@@ -293,13 +296,33 @@ def _compute_reply_to(self):
...
             if not mailing.mailing_model_name:
                 mailing.mailing_domain = ''
             else:
-                mailing.mailing_domain = repr(mailing._get_default_mailing_domain())
+                default_filter = repr(mailing._get_default_mailing_domain())
+                if mailing.mailing_favorite_filter_id:
+                    # If model is changed and filter is set, domain should reset to default one, emptying the filter,
+                    # or, if filter itself is changed (means model is same), simply apply filter's domain
+                    mailing.mailing_domain = mailing.mailing_favorite_filter_id.mailing_domain if mailing.mailing_model_name == mailing.mailing_favorite_filter_id.mailing_model_name else default_filter
+                else:
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/models/mailing.py:310]: Indeed its hackish. But we want to keep the domain of the last deleted filter. That means, user could change n number of filters, but we need to keep the domain from last filter, not the saved one we get from _origin.

File: addons/mass_mailing/models/mailing.py [hunk 17]
```diff
@@ -947,3 +970,20 @@ def _image_to_url(b64image: bytes):
...
+
+    name = fields.Char(string='Filter Name', required=True)
+    mailing_domain = fields.Char(string='Filter Domain', required=True)
+    mailing_model_name = fields.Char(string='Recipients Model Name', required=True, readonly=True)
+    mailing_model_description = fields.Char(string='Model', compute='_compute_mailing_model_description')
+
+    def _compute_mailing_model_description(self):
+        filter_models = self.env['ir.model'].search([('model', 'in', self.mapped('mailing_model_name'))])
+        for mailing_filter in self:
+            mailing_filter.mailing_model_description = filter_models.filtered(lambda x: x.model == mailing_filter.mailing_model_name).name
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/models/mailing.py:989]: Hello @awa-odoo 
Sure, will do that. I was just trying to avoid fetching in for loop.

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+    /**
+     * @override
+     * @private
+     */
+    _renderEdit() {
+        this._super.apply(this, arguments);
+        this._updateFilterIcons();
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:51]: Hello @awa-odoo 
Well, every time something is changed in edit mode, we need update the icons (the same way we update the external button on original m2o widget with _updateExternalButton method). Not sure how can we avoid it. Do you have any idea ?

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+    /**
+     * Creates a new favorite filter, with the name provided from drop-down and
+     * with the 'up to date' domain. If the input is blank, displays the warning
+     * and keeps the popup open by preventing event propagation.
+     *
+     * @private
+     * @param {event} ev
+     */
+    async _onFavoriteFilterSaveClick(ev) {
+        const filterInput = this.el.querySelector('input.o_mass_mailing_filter_name');
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:163]: Hello @awa-odoo 
As soon as the 'Save' button is clicked, the popup will be closed immediately.

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+            dataPointID: this.record.id,
+            changes: {
+                mailing_favorite_filter_id: {
+                    operation: 'DELETE',
+                    ids: [filterId]
+                },
+                mailing_domain: mailingDomain,
+            },
+        });
+        // prevent rage clicks to avoid possible breaks
```
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:150]: Done!

File: addons/mass_mailing/models/mailing.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/mass_mailing/models/mailing.py:308]: This is why we only reset the domain to "default_filter" if there is no filter_id AND if it was not just removed.
I was not suggesting to restore the one from _origin here, just to check if it's there to avoid resetting the filter.
Did you try it?

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:166]: Maybe add that as a note in the docstring?

File: addons/mass_mailing/models/mailing.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/mass_mailing/models/mailing.py:988]: I think it fetches the information from the cache so there is no database call involved.

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:51]: Ah yes, I see it's doing the same thing with "_updateExternalButton" indeed.
You can keep it like this then.

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (msh-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:39]: @awa-odoo Here we are not accessing the actual DOM element I think, we are accessing this.$el which may not be in DOM but it is still accessible in ```start``` method, @mub-odoo maybe you can try ```on_attach_callback``` method

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:166]: Changes Done!

File: addons/mass_mailing/models/mailing.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/models/mailing.py]: Changes Done !

File: addons/mass_mailing/models/mailing.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/models/mailing.py:308]: hello @awa-odoo 
i will try this,  but the domain is changed or different from last one 
and _origin gives last saved record in db and it returns None if 
record is first time created.

[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/mass_mailing/models/mailing.py:308]: In that case can you maybe do this:

```py
elif mailing.mailing_model_name != mailing._origin.mailing_model_name:
    # changed selected model, reset to default filter
    mailing.mailing_domain = default_filter
```

Make sure it works for every use cases of course.

I would really like to avoid that manual filter change in the JS.

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:39]: Yes I noticed that as well.
You can leave the code as is and add this to the existing docstring of the method:

`* see FieldMany2One#start for more details.`

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+     * the current state.
+     *
+     * @private
+     */
+    _updateFilterIcons: function () {
+        const hasDomain = this.recordData.mailing_domain && this.recordData.mailing_domain !== '[]';
+        // if the domain is empty, hide star icon that allows creating / deleting a filter
+        this.el.querySelector('.o_mass_mailing_filter_container').classList.toggle('invisible', !hasDomain);
+        // A filter can be saved if no filter is set, or if user empties the m2o input,
+        // (this.$input) or changes the value of m2o input (so that it's 'floating')
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:65]: I would put this in the docstring instead, such as:

```js
/**
* Updates the 'Add to favorite' / 'Remove' star icons' visibility based on
* the current state.
* 
* The filter can be saved if one of those conditions is matched:
* - No favorite filter is currently set
* - User just emptied the input
* - The input is currently being edited, known by the "this.floating" variable
*   see FieldMany2One#start for more information about the floating variable.
*
* @private
*/
```

**Review by awa-odoo — COMMENTED** [review]
Hello @mub-odoo 

Only that manual filter change thing left that I would like to solve.
Otherwise it looks good!

Don't forget to squash your "WIP" commit as well.

Thanks!

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:39]: done

File: addons/mass_mailing/models/mailing.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (mub-odoo) [reply addons/mass_mailing/models/mailing.py:308]: Hello @awa-odoo 
So did I, but could not find a proper way to deal with it in python :) Let me rephrase a bit again, to explain what we want. If you are creating a new record, there will be nothing in `_origin`. You could change the model as many times as you want, and you could create as many filters for any model as you want. But when you simply click on `star` icon to remove the last created filter, only the filter needs to be deleted but we need to keep the domain(note that we haven't even saved the record). How do we manage that with compute method? I think that all the conditions will lead to default filter being set, because there's no point of comparing anything with `_origin` state for new records. And it will be the case most of times in context of this task(people will play with domains / filters on new records).

[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/mass_mailing/models/mailing.py:308]: Yes I see what you mean.

I guess we can't help it then, it's not possible to do only in python.
Unless maybe we start playing with onchange but those are a bit deprecated anyway.

Thanks for trying and for keeping up with me!

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+    /**
+     * This widget now has two inputs, one for m2o and another for filter
+     * name. So this override makes sure that 'this.$input' now refers to
+     * only main input instead of both.
+     * see FieldMany2One#start for more details.
+     *
+     * @override
+     */
+    start() {
+        let def = this._super(...arguments);
```
Comment (std-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:52]: can you use `const` ?

File: addons/mass_mailing/static/tests/mass_mailing_favourite_filter_tests.js [hunk 23]
```diff
@@ -0,0 +1,284 @@
+/** @odoo-module alias=mass_mailing.FieldMassMailingFavoriteFilter.test */
+
+import FormView from 'web.FormView';
+import testUtils from 'web.test_utils';
+
+QUnit.module('mass_mailing_favourite_filter', {}, function () {
+QUnit.module('favorite filter widget', {
```
Comment (std-odoo) [inline addons/mass_mailing/static/tests/mass_mailing_favourite_filter_tests.js:7]: in this file you mix single and double quote, can you make it coherent ?

File: addons/mass_mailing/static/tests/mass_mailing_favourite_filter_tests.js [hunk 24]
```diff
@@ -0,0 +1,284 @@
...
+                    <field name="mailing_model_id"/>
+                    <field name="mailing_favorite_filter_domain" invisible="1"/>
+                    <field name="mailing_favorite_filter_id" widget="mailing_favorite_filter" options="{'no_create': '1', 'no_open': '1', 'domain_field': 'mailing_domain', 'model': 'mailing_model_id'}"/>
+                </form>`,
+            res_id: 1,
+            mockRPC: function (route, args) {
+                if (args.method === 'unlink' && args.model === 'mailing.favorite.filter') {
+                    assert.strictEqual(args.args[0], 1, "should pass correct filter ID for deletion");
+                }
+                if (args.method === 'write' && args.model === 'mailing.mailing') {
```
Comment (std-odoo) [inline addons/mass_mailing/static/tests/mass_mailing_favourite_filter_tests.js:180]:
```suggestion
                } else if (args.method === 'write' && args.model === 'mailing.mailing') {
```

File: addons/mass_mailing/tests/test_mailing_internals.py [hunk 25]
```diff
@@ -107,6 +107,47 @@ def test_mailing_computed_fields(self):
         self.assertEqual(mailing.reply_to_mode, 'update')
         self.assertFalse(mailing.reply_to)
 
+        # ---------- filter related computations ---------------
+
+        # prepare initial data
+        [favorite_filter_channel_1, favorite_filter_channel_2, favorite_filter_partner] = self.env['mailing.favorite.filter'].create([
```
Comment (std-odoo) [inline addons/mass_mailing/tests/test_mailing_internals.py:113]:
```suggestion

        favorite_filter_channel_1, favorite_filter_channel_2, favorite_filter_partner = self.env['mailing.favorite.filter'].create([
```

File: addons/mass_mailing/tests/test_mailing_internals.py [hunk 26]
```diff
@@ -107,6 +107,47 @@ def test_mailing_computed_fields(self):
...
+            {
+            'name': 'general channel',
```
Comment (std-odoo) [inline addons/mass_mailing/tests/test_mailing_internals.py:115]: indentation issue here

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 27]
```diff
@@ -0,0 +1,213 @@
...
+    * - User changed the domain when favorite filter is set
+    * - The input is currently being edited, known by the "this.floating" variable
+    *   (see FieldMany2One#start for more information about the floating variable).
+    *
+    * @private
+    */
+    _updateFilterIcons: function () {
+        const hasDomain = (this.recordData[this.domainFieldName] && this.recordData[this.domainFieldName] !== '[]');
+        this.el.querySelector('.o_mass_mailing_filter_container').classList.toggle('invisible', !hasDomain);
+        const canSaveFilter = !this.recordData.mailing_favorite_filter_id || !this.$input.val().length || this.floating || this.recordData.mailing_favorite_filter_domain !== this.recordData[this.domainFieldName];
```
Comment (std-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js:85]: can you split this line ?

File: addons/mass_mailing/static/tests/mass_mailing_favourite_filter_tests.js [hunk 28]
```diff
@@ -0,0 +1,284 @@
...
+            model: 'mailing.mailing',
+            data: this.data,
+            arch: `<form>
+                    <field name="display_name"/>
+                    <field name="subject"/>
+                    <field name="mailing_domain"/>
+                    <field name="mailing_model_name" invisible="1"/>
+                    <field name="mailing_model_id"/>
+                    <field name="mailing_favorite_filter_domain" invisible="1"/>
+                    <field name="mailing_favorite_filter_id" widget="mailing_favorite_filter" options="{'no_create': '1', 'no_open': '1', 'domain_field': 'mailing_domain', 'model': 'mailing_model_id'}"/>
```
Comment (std-odoo) [inline addons/mass_mailing/static/tests/mass_mailing_favourite_filter_tests.js:115]: can you split this line ?

**Review by std-odoo — COMMENTED** [review]
Hi :)

Very cool feature :D I added some notes (mainly style, code is globally good :) )

**Review by tde-banana-odoo — COMMENTED** [review]
Review! Thanks for your PR, looks clean and interesting. Code is compact, localized and correctly structured. Good job :) .

I added some naming and styling comments, just to simplify naming or make it coherent. We should try to merge it for the upcoming freeze, so maybe just ask to Lna and me to re-check a bit post-changes, so that we speedup the merge :) .

Thanks !

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 29]
```diff
@@ -0,0 +1,215 @@
...
+ */
+const MassMailingFavoriteFilter = FieldMany2One.extend({
+    template: 'mass_mailing.favorite_filter',
+    events: Object.assign({}, FieldMany2One.prototype.events, {
+        'click .o_mass_mailing_btn_save_filter': '_onFavoriteFilterSaveClick',
+        'click .o_mass_mailing_remove_filter': '_onFavoriteFilterRemoveClick',
+        'keydown .o_mass_mailing_filter_name': '_onFavoriteFilterNameInputKeydown',
+        'shown.bs.dropdown': '_onFavoriteFilterDropdownShown',
+    }),
+    resetOnAnyFieldChange: true,
```
[the remark this answers is not in the document]
Comment (dja-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js]: It's added on purpose, to get the latest `recordData` (reflecting live user changes and not the previously saved) which we need in widget.

File: addons/mass_mailing/static/src/scss/mass_mailing_mobile.scss [hunk 30]
```diff
@@ -4,4 +4,7 @@
         margin: 0 auto;
         justify-content: space-between !important;
     }
-}
\ No newline at end of file
+    .o_form_editable div[name="mailing_favorite_filter_id"] {
+        width: 100% !important;
```
[the remark this answers is not in the document]
Comment (dja-odoo) [reply addons/mass_mailing/static/src/scss/mass_mailing_mobile.scss:8]: We've used `w-auto` bootstrap class in the view which already uses important. And we need to override that for small devices.

File: addons/mass_mailing/static/src/scss/mass_mailing.scss [hunk 31]
```diff
@@ -115,6 +115,13 @@
     .o_sms_container ~ .o_mail_add_emoji{
         bottom: 55px;
     }
+    // Favorite filter
+    .o_mass_mailing_filter_container {
```
[the remark this answers is not in the document]
Comment (dja-odoo) [reply addons/mass_mailing/static/src/scss/mass_mailing.scss]: Done

File: addons/mass_mailing/static/src/xml/mass_mailing.xml [hunk 32]
```diff
@@ -15,4 +15,27 @@
             </a>
         </t>
     </div>
+
+    <!-- Template for 'mailing_favorite_filter' widget, specific for mailing form view -->
+    <t t-name="mass_mailing.favorite_filter" t-extend="FieldMany2One" primary="True">
```
[the remark this answers is not in the document]
Comment (dja-odoo) [reply addons/mass_mailing/static/src/xml/mass_mailing.xml]: Done

[the remark this answers is not in the document]
Comment (dja-odoo) [reply addons/mass_mailing/static/src/xml/mass_mailing.xml]: Done

File: addons/mass_mailing/models/mailing_favorite_filter.py [hunk 33]
```diff
@@ -0,0 +1,17 @@
...
+
+from odoo import fields, models
+
+
+class MailingFavoriteFilter(models.Model):
+    """ This model stores mass mailing domain as filters (quite similar to ir.filters but
+    dedicated to mass mailing), so that frequent domains can be reused easily.
+    """
+    _name = 'mailing.favorite.filter'
+    _description = 'Saved filters for Mass Mailing'
```
[the remark this answers is not in the document]
Comment (dja-odoo) [reply addons/mass_mailing/models/mailing_favorite_filter.py]: Done

File: addons/mass_mailing/models/mailing.py [hunk 34]
```diff
@@ -309,13 +312,17 @@ def _compute_reply_to(self):
...
-    @api.depends('mailing_model_name', 'contact_list_ids')
+    @api.depends('mailing_model_name', 'contact_list_ids', 'mailing_favorite_filter_id')
     def _compute_mailing_domain(self):
         for mailing in self:
             if not mailing.mailing_model_name:
                 mailing.mailing_domain = ''
             else:
-                mailing.mailing_domain = repr(mailing._get_default_mailing_domain())
+                default_filter = repr(mailing._get_default_mailing_domain())
+                if mailing.mailing_model_name == mailing.mailing_favorite_filter_id.mailing_model_name:
```
[the remark this answers is not in the document]
Comment (dja-odoo) [reply addons/mass_mailing/models/mailing.py]: Done

File: addons/mass_mailing/models/mailing.py [hunk 35]
```diff
@@ -125,6 +125,9 @@ def _get_default_mail_server_id(self):
     mailing_domain = fields.Char(
         string='Domain', compute='_compute_mailing_domain',
         readonly=False, store=True)
+    mailing_favorite_filter_id = fields.Many2one('mailing.favorite.filter', string='Favorite Filter',
```
[the remark this answers is not in the document]
Comment (dja-odoo) [reply addons/mass_mailing/models/mailing.py]: Done

File: addons/mass_mailing/security/ir.model.access.csv [hunk 36]
```diff
@@ -21,3 +21,4 @@ access_mailing_list_merge,access.mailing.list.merge,model_mailing_list_merge,mas
 access_mailing_mailing_test,access.mailing.mailing.test,model_mailing_mailing_test,mass_mailing.group_mass_mailing_user,1,1,1,0
 access_mailing_contact_to_list,access.mailing.contact.to.list,model_mailing_contact_to_list,mass_mailing.group_mass_mailing_user,1,1,1,1
 access_mailing_schedule_date,access.mailing.schedule.date,model_mailing_mailing_schedule_date,mass_mailing.group_mass_mailing_user,1,1,1,1
+access_mail_filters_mm_user,access.mailing.filters,model_mailing_favorite_filter,mass_mailing.group_mass_mailing_user,1,1,1,1
```
[the remark this answers is not in the document]
Comment (dja-odoo) [reply addons/mass_mailing/security/ir.model.access.csv]: Done

File: addons/mass_mailing/models/mailing_favorite_filter.py [hunk 37]
```diff
@@ -0,0 +1,17 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+
+
+class MailingFavoriteFilter(models.Model):
+    """ This model stores mass mailing domain as filters (quite similar to ir.filters but
+    dedicated to mass mailing), so that frequent domains can be reused easily.
+    """
+    _name = 'mailing.favorite.filter'
```
[the remark this answers is not in the document]
Comment (dja-odoo) [reply addons/mass_mailing/models/mailing_favorite_filter.py]: Done 😄

File: addons/mass_mailing/static/src/scss/mass_mailing_mobile.scss (hunk 30 above)
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mass_mailing/static/src/scss/mass_mailing_mobile.scss:8]: Ok (worth a small comment in css code maybe ? )

File: addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js [hunk 38]
```diff
@@ -0,0 +1,215 @@
+/** @odoo-module alias=mass_mailing.MassMailingFavoriteFilter **/
```
[the remark this answers is not in the document]
Comment (dja-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_favorite_filter_widget.js]: Done

File: addons/mass_mailing/static/src/scss/mass_mailing_mobile.scss (hunk 30 above)
[the remark this answers is not in the document]
Comment (dja-odoo) [reply addons/mass_mailing/static/src/scss/mass_mailing_mobile.scss:10]: Done

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-ff
